File: survey/survey-domain/src/main/java/me/nalab/survey/domain/feedback/Bookmark.java
package me.nalab.survey.domain.feedback;

import java.time.Instant;
import java.util.Objects;

import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

@Builder
@Getter
@ToString
@EqualsAndHashCode
public class Bookmark {

	private static final boolean BOOKMARK_DEFAULT_STATE = false;

	private boolean isBookmarked = BOOKMARK_DEFAULT_STATE;
	private Instant bookmarkedAt;

	public Bookmark(boolean isBookmarked, Instant bookmarkedAt) {
		Objects.requireNonNull(bookmarkedAt, () -> "Null value on new Bookmark(..., bookmarkedAt)");
		this.isBookmarked = isBookmarked;
		this.bookmarkedAt = bookmarkedAt;
	}

	public void replaceIsBookmarked() {
		this.isBookmarked = !isBookmarked;
		this.bookmarkedAt = Instant.now();
	}

}

File: src/main/kotlin/me/nalab/api/survey/domain/feedback/Bookmark.kt
package me.nalab.api.survey.domain.feedback

import me.nalab.core.time.TimeUtil
import java.time.Instant
import javax.persistence.Column
import javax.persistence.Embeddable

@Embeddable
class Bookmark(
    @Column(name = "is_bookmarked", nullable = false)
    var isBookmarked: Boolean = BOOKMARK_DEFAULT_STATE,

    @Column(name = "bookmarked_at", columnDefinition = "TIMESTAMP(6)", nullable = false)
    var bookmarkedAt: Instant,
) {

    companion object {
        private const val BOOKMARK_DEFAULT_STATE = false

        fun impossible(): Bookmark = Bookmark(bookmarkedAt = TimeUtil.toInstant())
    }
}
